community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java

```diff
@@ -0,0 +1,385 @@
+            // Settings that were updated by our pass-through.
+            Set<String> redefined = new HashSet<>();
+            // Skip settings that are either already defined in the conf file, or were defined earlier in our pass.
+            Predicate<String> skip = s -> existingSettings.containsKey( s ) || redefined.contains( s );
+            try ( Stream<String> lines = Files.lines( confFile, StandardCharsets.UTF_8 ) )
+            {
+                lines.forEachOrdered( line ->
+                {
+                    line = tryUpdateSetting( line, skip, redefined, overwrite, initialHeapSize.name(), heap );
+                    line = tryUpdateSetting( line, skip, redefined, overwrite, maxHeapSize.name(), heap );
+                    line = tryUpdateSetting( line, skip, redefined, overwrite, pagecache_memory.name(), pagecache );
+                    out.accept( line );
+                } );
+            }
+            defineIfMissing( out, redefined, initialHeapSize.name(), heap );
+            defineIfMissing( out, redefined, maxHeapSize.name(), heap );
+            defineIfMissing( out, redefined, pagecache_memory.name(), pagecache );
+        }
+        catch ( Exception e ) // Also catch runtime exceptions due to use of Streams.
+        {
+            throw new CommandFailed( "Failed to read configuration file: " + confFile, e );
+        }
+    }
+
+    private String tryUpdateSetting( String line, Predicate<String> alreadyDefined, Set<String> redefined,
+                                     boolean overwrite, String name, String value )
+    {
+        int settingIndex = line.indexOf( name );
+        if ( settingIndex == -1 )
+        {
+            // The setting we're looking for is not present in this line, so we just return it unmodified.
+            return line;
+        }
+        // Otherwise, we need to modify this line, such that it reflects the new value.
+        // Except, if 'overwrite' is false, then we don't modify the line if the setting is not commented out.
+        int firstComment = line.indexOf( '#' );
+        if ( !overwrite && (settingIndex < firstComment || firstComment == -1 || alreadyDefined.test( name )) )
+        {
+            // The existing setting is enabled, but we are not allowed to overwrite it, so we return the line
+            // unmodified.
+            return line;
+        }
+        // We also need to make sure that what we have found, is not just a mention of the setting in some comment.
+        // We recognise commented out settings (as opposed to mentions of the setting in a comment) by the equals sign
+        // that follows the setting name, and the absence of any prose text before the setting name.
+        int endOfSettingIndex = settingIndex + name.length();
+        int equalsSign = line.indexOf( '=', endOfSettingIndex );
+        if ( equalsSign == -1 || containsProse( line, endOfSettingIndex, equalsSign ) ||
```

Comment:
Why do you differentiate between different kinds of comments?

Reply:
In `neo4j.conf` we have a passage that look like this:

```# The legacy policy is a special policy which is not defined in
# the policy configuration section, but rather derives from
# dbms.directories.certificates and associated files
# (by default: neo4j.key and neo4j.cert). Its use will be deprecated.```

If a similar passage gets introduced by us or by an admin, then we should avoid clobbering it.

Reply:
well sure, but we should just avoid modifying comments period. So I don't see the reason why we need to differentiate between different kinds of comments. Just ignore all lines where first non-whitespace character is `#`.

Reply:
And then instead just append the missing settings to the end of the file?